File: src/jax/java/jaxdemo/eventhandler/QueryEventHandler.java
package jax.java.jaxdemo.eventhandler;

import java.util.ArrayList;
import java.util.List;

import jax.java.jaxdemo.events.QueryDataEvent;
import jax.java.jaxdemo.view.PersonRecord;
import jax.java.jaxdemo.view.PersonRecordView;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class QueryEventHandler implements EventHandler<QueryDataEvent> {

	private PersonRecordView personRecordView;
	private Logger log = LogManager.getLogger(this);
	private static int MAX_PRINT_RECORDS = 5;

	public QueryEventHandler(PersonRecordView personRecordView) {
		this.personRecordView = personRecordView;
	}

	@Override
	public void applyEvent(QueryDataEvent event) throws Exception {
		List<PersonRecord> queryResults = new ArrayList<PersonRecord>();

		if (event.getQuerySsn() != null) {
			PersonRecord queryBySsnResults = personRecordView.getSsnIndex().get(event.getQuerySsn());
			if (queryBySsnResults != null) {
				queryResults.add(queryBySsnResults);
			}
		} else if (event.getQueryDate() != null) {
			List<PersonRecord> queryByDateResults = personRecordView.getDateIndex().get(event.getQueryDate());
			if (queryByDateResults != null) {
				queryResults.addAll(queryByDateResults);
			}
		} else {
			queryResults.addAll(personRecordView.getViewStore().values());
		}

		long queryTime = System.currentTimeMillis() - event.getQueryStartTime();
		log.info("Completed Query ssn=" + event.getQuerySsn() + " dob=" + event.getQueryDate() + " in " + queryTime + "ms (" + queryResults.size()
				+ " records)");

		int count = 0;
		for (PersonRecord p : queryResults) {
			count++;
			log.info(p.toString());
			if (count >= MAX_PRINT_RECORDS) {
				log.info((queryResults.size() - count) + " More Records...");
				break;
			}
		}

	}

}

File: src/jax/scala/jaxdemo/eventhandler/QueryEventHandler.scala
package jax.scala.jaxdemo.eventhandler

import jax.scala.jaxdemo.view.PersonRecordView
import org.apache.logging.log4j.LogManager
import jax.scala.jaxdemo.view.PersonRecord
import scala.collection.mutable.ListBuffer
import jax.scala.jaxdemo.events.QueryDataEvent

case class QueryEventHandler(personRecordView: PersonRecordView) extends EventHandler[QueryDataEvent] {

  val log = LogManager.getLogger(this)
  val MAX_PRINT_RECORDS = 5;

  override def applyEvent(event: QueryDataEvent) {
    val results: List[PersonRecord] =
      if (event.querySsn != null && personRecordView.ssnIndex.contains(event.querySsn))
        personRecordView.ssnIndex.get(event.querySsn).get :: Nil
      else if (event.queryDate != null)
        personRecordView.dateIndex.getOrElse(event.queryDate, new ListBuffer[PersonRecord]).toList
      else
        personRecordView.viewStore.values.toList

    val queryTime = System.currentTimeMillis() - event.queryStartTime
    log.info(s"Completed Query ssn=${event.querySsn} dob=${event.queryDate} in $queryTime ms (${results.size} records)");
    logResults(results)

  }

  private def logResults(results: List[PersonRecord]) {

    for (i <- 0 until MAX_PRINT_RECORDS; if i < results.size) {
      log.info(results(i));
    }
    if (MAX_PRINT_RECORDS < results.size)
      log.info((results.size - MAX_PRINT_RECORDS) + " More Records...")
  }
}